pymc_marketing/clv/models/basic.py

```diff
@@ -49,21 +43,87 @@ def fit(self, fit_method="mcmc", **kwargs):
         kwargs:
             Other keyword arguments passed to the underlying PyMC routines
         """
+
+        self.build_model()
+
         if fit_method == "mcmc":
-            res = self._fit_mcmc(**kwargs)
+            self._fit_mcmc(**kwargs)
         elif fit_method == "map":
-            res = self._fit_MAP(**kwargs)
+            self._fit_MAP(**kwargs)
         else:
             raise ValueError(
                 f"Fit method options are ['mcmc', 'map'], got: {fit_method}"
             )
-        self.fit_result = res
-        return res
 
-    def _fit_mcmc(self, **kwargs):
-        """Draw samples from model posterior using MCMC sampling"""
+        with warnings.catch_warnings():
+            warnings.filterwarnings(
+                "ignore",
+                category=UserWarning,
+                message="The group fit_data is not defined in the InferenceData scheme",
+            )
+            self.idata.add_groups(fit_data=self.data.to_xarray())  # type: ignore
+
+        return self.idata
+
+    def _fit_mcmc(self, **kwargs) -> az.InferenceData:
+        """
+        Fit a model using the data passed as a parameter.
+        Sets attrs to inference data of the model.
+
+
+        Parameters
+        ----------
+        X : array-like if sklearn is available, otherwise array, shape (n_obs, n_features)
+            The training input samples.
+        y : array-like if sklearn is available, otherwise array, shape (n_obs,)
+            The target values (real numbers).
+        **kwargs : Any
+            Custom sampler settings can be provided in form of keyword arguments.
+
+        Returns
+        -------
+        self : az.InferenceData
+            returns inference data of the fitted model.
+        """
+        sampler_config = {}
+        if self.sampler_config is not None:
+            sampler_config = self.sampler_config.copy()
+        sampler_config.update(**kwargs)
+        self.idata = self.sample_model(**sampler_config)
+        return self.idata
+
+    def sample_model(self, **kwargs):
+        """
+        Sample from the PyMC model.
+
+        Parameters
+        ----------
+        **kwargs : dict
+            Additional keyword arguments to pass to the PyMC sampler.
+
+        Returns
+        -------
+        xarray.Dataset
+            The PyMC samples dataset.
+
+        Raises
+        ------
+        RuntimeError
+            If the PyMC model hasn't been built yet.
+
+        """
+        if self.model is None:
+            raise RuntimeError(
+                "The model hasn't been built yet, call .build_model() first or call .fit() instead."
+            )
+
         with self.model:
-            return pm.sample(**kwargs)
+            sampler_args = {**self.sampler_config, **kwargs}
+            idata = pm.sample(**sampler_args)
+            idata.extend(pm.sample_posterior_predictive(idata))
```

Comment:
We can't call `sample_posterior_predictive` by default, for the same reason we couldn't call `sample_prior_predictive`. Not all CLV models have a standard likelihood that can be sampled from.

Is this `sample` something that in the `ModelBuilder` class calls all three sample methods in succession? If so we should just raise NotImplementedError in the base CLV class

Reply:
we're anyway already re-defining the sample function inside basic.py so I'd say let's remove the sample_posterior_predictive from default sampling